### PR TITLE
Altair transition tests

### PR DESCRIPTION
**Pull request by tersec**
- add phase 0 to Altair transition tests.
- both `upgrade_to_altair()` callers would prefer a reference, so construct one directly rather than rely on `newClone()`.
- add a `ForkedHashedBeaconState` type which doesn't have DAG content, but does track which fork (Phase 0, Altair, or in the future, merge) that `BeaconState` is on. A next step here is to replace the `HashedBeaconState` in `StateData` with `ForkedHashedBeaconState`, so that the in-place modifications that `state_transition(var Foo)` and `state_transition_block(var Foo)` can be kept.
- leaves in place the one remaining `slotProcessed`-based caller, in `applyBlock()` in `blockchain_dag`.
- remove `test_state_transition`, which is mostly a smoke test, checking a couple of basic sanity conditions. However, the official EF test vectors which `test_fixture_transition` and `test_fixture_sanity_block` use are more comprehensive and rigorous.

### CODE REVIEW

File: beacon_chain/spec/state_transition.nim [hunk 1]
```diff
@@ -264,6 +268,169 @@ proc state_transition*(
 
   true
 
+# TODO this will become the main StateData type, but isolate the change to
```
Comment (arnetheduck) [inline beacon_chain/spec/state_transition.nim]: the thing to remember here is that `StateData` exists to tie the low-level spec code with `BlockRef`, essentially - in the type system, `StateData` is separated because while the spec works with roots, it doesn't have a notion of a DAG at all - it's not entirely correct to subvert it to become a place where case objects happen.

Basically, state_transition is oblivious to the dag, except for the most immediate parent - `StateData.blck` basically points to which was the latest block that was applied to the state and it's the responsibility of chaindag to maintain this invariant. 

The fork stuff will likely need a _new_ type for the case object at some level or it would have to import the entire DAG interface which would be wrong.

File: beacon_chain/spec/state_transition.nim [hunk 2]
```diff
@@ -264,6 +268,169 @@ proc state_transition*(
...
+  of forkAltair: hash_tree_root(x.hbsAltair.data)
+
+template callWithBS*(op: untyped, y: StateDataStaging): untyped =
+  let bs {.inject.} =
+    case y.beaconStateFork:
+    of forkPhase0: y.hbsPhase0.data
+    of forkAltair: y.hbsAltair.data
+  op
+
+proc state_transition*(
```
Comment (arnetheduck) [inline beacon_chain/spec/state_transition.nim:392]: I'd prefer to see separate state transition functions for altair and phase0 - basically, `state_transition` works with either one beaconstate kind or the other, but never both - there's a special function for making that transition in particular - this means that state_transition itself doesn't ever need access to _both_ case kinds for any kind of conversion logic.

There's also a reason why `state_transition` takes `HashedBeaconState` - this is to show that it's responsible for updating the beaconstate and the root, but _not_ the `blck` field of `BeaconState` - again, this is to separate the dag from the state transtion and we should maintain this distinction.

Comment (tersec) [reply beacon_chain/spec/state_transition.nim:392]: I'm not sure that's possible.

https://github.com/ethereum/eth2.0-specs/blob/v1.1.0-alpha.6/specs/altair/fork.md#upgrading-the-state:
> The upgrade occurs after the completion of the inner loop of `process_slots` that sets state.slot equal to `ALTAIR_FORK_EPOCH * SLOTS_PER_EPOCH`. Care must be taken when transitioning through the fork boundary as implementations will need a modified state transition function that deviates from the Phase 0 document. In particular, the outer `state_transition` function defined in the Phase 0 document will not expose the precise fork slot to execute the upgrade in the presence of skipped slots at the fork boundary. Instead the logic must be within `process_slots`.

Since `process_slots` is within the dynamic scope, typically, of `state_transition` (and is here too, though it's inlined so it's not labeled as such), `state_transition` seems like it's going to have to be involved.

Comment (arnetheduck) [reply beacon_chain/spec/state_transition.nim:392]: the point is a little bit that we can leave the logic for making the switch between one kind of state and another outside of the "core" state transition - in part, this is because we often process blocks and slots separately in chaindag - we don't call state_transition to make the slot transition, actually, because we need to capture the state _without_ the block applied - state_transition as a function has the wrong call signature, in many ways, to support chaindag.

Comment (tersec) [reply beacon_chain/spec/state_transition.nim:392]: That said, I agree, that `StateData` is a DAG object, and that's a problem in some use cases, and in particular, `state_transition` shouldn't be operating on DAG objects.

One approach is since `process_slots()` is the first part of `state_transition`, split `state_transition` from:
```python
def state_transition(state: BeaconState, signed_block: SignedBeaconBlock, validate_result: bool=True) -> None:
    block = signed_block.message
    # Process slots (including those with no blocks) since block
    process_slots(state, block.slot)
    # Verify signature
    if validate_result:
        assert verify_block_signature(state, signed_block)
    # Process block
    process_block(state, block)
    # Verify state root
    if validate_result:
        assert block.state_root == hash_tree_root(state)
```
into
```python
def state_transition_part1(state: BeaconState, signed_block: SignedBeaconBlock, validate_result: bool=True) -> None:
    block = signed_block.message
    # Process slots (including those with no blocks) since block
    process_slots(state, block.slot)
```
and
```python
def state_transition_part2(state: BeaconState, signed_block: SignedBeaconBlock, validate_result: bool=True) -> None:
    block = signed_block.message
    # Verify signature
    if validate_result:
        assert verify_block_signature(state, signed_block)
    # Process block
    process_block(state, block)
    # Verify state root
    if validate_result:
        assert block.state_root == hash_tree_root(state)
```

Where `state_transition_part1` deals with switching, and `state_transition_part2` is per-`BeaconState`-type.

However, the issue remains for the `state_transition` interface, for the places which want a spec-like `state_transition`: one can call it on a `(phase 0 state, phase 0 block)` and somehow have to indicate, sometimes, that one gets an (altair state) back.

Comment (arnetheduck) [reply beacon_chain/spec/state_transition.nim:392]: put another way, in chaindag we basically call `process_slots` one by one (because the post-slot, pre-block state is used sometimes) until we arrive at the right slot, then we _effectively_ call `process_block`, invoking it using `state_transition` as a wrapper basically so it can be grepped for - the `state_transition` we call never does slot processing (`UpdateFlags.slotProcessed`) - I expect that both for the sake of the database and for most of the other code, the logic that chooses the fork moment and everything that surrounds it will likely end up living in `chaindag` more than here.

Comment (tersec) [reply beacon_chain/spec/state_transition.nim:392]: Sure, to the extent that the DAG doesn't actually want `state_transition()` per se, then it's avoidable.

I've been careful to avoid conceptual redesigns of things, as much as feasible. Almost everything has been a "dumb", mechanical refactoring throughout this Altair readiness preparation, and that's intentional. It might be worth deviating from that here, but I'd want to consider it more.

Comment (arnetheduck) [reply beacon_chain/spec/state_transition.nim:392]: it would actually be more honest to call `process_block` directly in chaindag - the only reason that's not done right now is an optimization for `HashedBeaconState.root` - when we _know_ there's a block coming, we can pluck the root from the block instead of calling `htr` - there are probably prettier ways to express that optimization however

Comment (tersec) [reply beacon_chain/spec/state_transition.nim:392]: That's my inclination -- and my point about `state_transition()` holds. That interface, however useful or not it is for `blockchain_dag`, just fundamentally has to know how to do the transition. We can choose not to support it, e.g., by setting up `test_fixture_transition` to do the same kind of thing you're proposing for the DAG, and just stipulating that the various other `state_transition` calls (e.g., in the fuzzers) don't transition at all.

This wouldn't work for `ncli_db`, though.

This is sort of what I mean -- dropping this capability is quite nontrivial across the codebase, since not only `consensus_object_pools` uses `state_transition`, and other users don't necessarily have the same manual approach to `process_slots()` then `state_transition()`-as-`process_block()`.

File: beacon_chain/spec/state_transition.nim (hunk 1 above)
Comment (tersec) [reply beacon_chain/spec/state_transition.nim]: I basically agree; the main consideration is pre- or post- first testnet.

This was a design blunder in using `StateData` (rather than introducing another type). If it's timeline-feasible to switch to something which just has the case object, but doesn't have any DAG content, then, great.

Since state transition function, including tests, are basically complete, this is something I'll look at in detail next.

File: beacon_chain/spec/state_transition.nim (hunk 2 above)
Comment (tersec) [reply beacon_chain/spec/state_transition.nim:311]: Also,

> The upgrade occurs after the completion of the inner loop of `process_slots`

Seems to be false. I tried it, and it doesn't work, unless one splits the inner loop of `process_slots` in a way effectively equivalent to what's done here. There's a test case with no blocks until slot 96 and the `fork_epoch` set to 2. Empirically, the test case wants the transition to occur at slot 64, not 96. That's inconsistent with the spec wording.

**Comment by arnetheduck** [comment]
> other users don't necessarily have the same manual approach

it would perhaps be better if they did so we use one approach across everywhere - since the "split" approach is necessary regardless, we should probably come up with an API that serves that use case well and leave the combined `state_transition` behind.

To not regress on performance, the important part is that when replaying, one should never call hash_tree_root except for empty slots - the root is needed to fill in `state.state_roots` but can "normally" come from the block field instead of being computed.

When validating the block, we compute it when validating, but then use the root from the block regardless which makes the flow easier.

**Comment by tersec** [comment]
These are the current `unstable` `state_transition()` callers:
```
beacon_chain/consensus_object_pools/blockchain_dag.nim(729, 12)
beacon_chain/consensus_object_pools/block_clearance.nim(179, 10)
nbench/scenarios.nim(165, 19)
ncli/ncli_db.nim(210, 14)
ncli/ncli_db.nim(532, 12)
ncli/ncli_db.nim(759, 12)
ncli/ncli.nim(89, 10)
nfuzz/libnfuzz.nim(121, 7)
tests/official/phase0/test_fixture_sanity_blocks.nim(53, 25)
tests/official/phase0/test_fixture_sanity_blocks.nim(58, 25)
tests/test_state_transition.nim(116, 11)
tests/test_state_transition.nim(59, 20)
tests/test_state_transition.nim(79, 22)
```

Of these:

- `blockchain_dag.nim` and `block_clearance.nim` are the core places where the `process_block()`-alone version either is already effectively used or should be.

- `libnfuzz`, `ncli.nim`, and `nbench_scenarios` straightforwardly export the spec `state_transition()` interface to some variety of end-user accessibility. These, either we have something that's equivalent to the spec `state_transition()`, those would have to re-invent, or we'd remove them (potentially in favor of something else).

- `test_state_transition.nim` is epiphenomenal to `state_transition()`. It's spec-related, but the spec has other tests for `state_transition()` (finality and block sanity, and in Altair the transition tests). More test coverage is better, but if these get in the way, they could be dropped temporarily or permanently without much loss given the EF test vector spec/code coverage.

- `test_fixture_sanity_blocks` takes advantage of the `process_slots()` part of `state_transition()`, but it could use the manual approach easily enough. It's a small adjustment either way. It doesn't, however, need or want a DAG. Those tests in themselves don't ever switch fork, so it's okay for that part to be dropped. The transition tests do.

- This leaves the three `ncli_db` usages. They're in `cmdBench()`, `cmdValidatorPerf()`, and `cmdValidatorDb()`, the `bench`, `validatorPerf`, and `validatorDb` commands respectively.

The edge cases are really the transition tests and the `ncli_db` ones: they don't have, need, or want a DAG, but they do need to transition between forks. If it were just the transition tests, I'd say put those as a one-off special case in the test code, but three places in `ncli_db` is enough to warrant something more systematic. It does mean that the DAG alone can't be the only place which knows about fork transitions though.

`libnfuzz`, `ncli.nim`, and `nbench_scenarios` present a different question: if `state_transition` is a end-user interface, then it still at some level needs to be supported, and it needs to be supported in a way that, yes, understands fork transitions without any DAG. However, they're fundamentally optional.

`ncli_db` users have the opposite issue: while they also need to understand fork transitions without a DAG, they're less optional, but also not directly visible to end-users, so there's more flexibility in architecture and implementation.

**Comment by arnetheduck** [comment]
Ideally we can develop a good public interface for the block part, then update the existing state_transition to simply call that - this should solve the conundrum without making too many architectural sacrifices - the key issue right now I think is that we use `state_transittion` + flag when we mean "just process the block".

After that change, formulating a `state_transition` function that works for the utilities and tests will be easier, but above   all will ensure that the dag and tests do something .. similar at least.

File: beacon_chain/spec/state_transition.nim [hunk 3]
```diff
@@ -264,6 +268,169 @@ proc state_transition*(
...
+  of forkPhase0: x.hbsPhase0.root
+  of forkAltair: x.hbsAltair.root
+
+template hash_tree_root*(x: StateDataStaging): Eth2Digest =
+  case x.beaconStateFork:
+  of forkPhase0: hash_tree_root(x.hbsPhase0.data)
+  of forkAltair: hash_tree_root(x.hbsAltair.data)
+
+template callWithBS*(op: untyped, y: StateDataStaging): untyped =
+  let bs {.inject.} =
```
Comment (zah) [inline beacon_chain/spec/state_transition.nim]: This can be a template. A copy is expensive and doesn't work for `var` parameters.

**Comment by tersec** [comment]
> Ideally we can develop a good public interface for the block part, then update the existing state_transition to simply call that - this should solve the conundrum without making too many architectural sacrifices - the key issue right now I think is that we use `state_transittion` + flag when we mean "just process the block".
> 
> After that change, formulating a `state_transition` function that works for the utilities and tests will be easier, but above all will ensure that the dag and tests do something .. similar at least.

https://github.com/status-im/nimbus-eth2/blob/505794f4303d8c9194224aa9962150fa070e758b/beacon_chain/spec/state_transition.nim#L221-L256 shows one version of this: it splits `state_transition()` into `state_transition_slots()` and `state_transition_block()`, and code which only needs the latter can skip the former.

Since `slotProcessed` implies the beaconblock's slot is equal to the state slot, the five callers which set `slotProcessed` (the three `ncli_db` ones, the `block_clearance.nim` one, and the `blockchain_dag.nim` would in theory be drop-in replaceable with this `state_transition_block()`.

File: beacon_chain/spec/state_transition.nim (hunk 3 above)
Comment (tersec) [reply beacon_chain/spec/state_transition.nim]: I'm confused -- they are all templates already?

File: beacon_chain/spec/state_transition.nim (hunk 1 above)
Comment (tersec) [reply beacon_chain/spec/state_transition.nim]: I added a `ForkedHashedBeaconState` type for this purpose (naming tentative).

**Review by arnetheduck — APPROVED** [review]
